### PR TITLE
New fatigue example

### LINKED ISSUE #313: Add Blade fatigue example
See Javier Viques example.

### PR DESCRIPTION
**Pull request by janvonrickenbach**
closes #313 

Based on #396 because I first tried to implement it in terms of the function implemented there. Turns out this does not work because of a bug in dpf core (see comment).

### CODE REVIEW

File: examples/006_filter_composite_data_example.py [hunk 1]
```diff
@@ -4,9 +4,15 @@
 Filter result data by different criteria
 ----------------------------------------
 
-This example show how data filtering can be used for custom postprocessing of
+This example shows how data filtering can be used for custom postprocessing of
```
Comment (PipKat) [inline examples/006_filter_composite_data_example.py:7]:
```suggestion
This example shows how you can use data filtering for custom postprocessing of
```
Use active voice when possible.

File: examples/006_filter_composite_data_example.py [hunk 2]
```diff
@@ -4,9 +4,15 @@
...
 layered composites. You can filter strains and stresses by material, layer, or
-analysis ply. The example filters data by layer, spot, and node, as well as material
+analysis ply. Filtering by analysis ply is implemented on the server side and
+exposed with the function :func:`.get_ply_wise_data`. In this case the data is
```
Comment (PipKat) [inline examples/006_filter_composite_data_example.py:10]:
```suggestion
exposed with the :func:`.get_ply_wise_data` function. In this case, the data is
```
Google dev style guide says to put the noun describing a tagged entity after the entity.

File: examples/006_filter_composite_data_example.py [hunk 3]
```diff
@@ -4,9 +4,15 @@
...
+filtered (and reduced) on the server side and only the resulting field is returned
```
Comment (PipKat) [inline examples/006_filter_composite_data_example.py:11]:
```suggestion
filtered (and reduced) on the server side, and only the resulting field is returned
```

File: examples/006_filter_composite_data_example.py [hunk 4]
```diff
@@ -4,9 +4,15 @@
...
+to the client. This is the recommended way to filter data if possible.
+For more complex filtering, the data is transferred to the client and side and filtered
+using numpy functionality.
```
Comment (PipKat) [inline examples/006_filter_composite_data_example.py:14]:
```suggestion
For more complex filtering, the data is transferred to the client side and filtered
using NumPy functionality.
```

File: examples/006_filter_composite_data_example.py [hunk 5]
```diff
@@ -55,16 +58,52 @@
...
+# Filter data by analysis ply
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+# %%
+# List all available analysis plies.
+all_ply_names = get_all_analysis_ply_names(composite_model.get_mesh())
+all_ply_names
+
+# %%
+# The easiest way to filter data by analysis ply is to use the function :func:`.get_ply_wise_data`.
```
Comment (PipKat) [inline examples/006_filter_composite_data_example.py:71]:
```suggestion
# The easiest way to filter data by analysis ply is to use the :func:`.get_ply_wise_data` function.
```

File: examples/006_filter_composite_data_example.py [hunk 6]
```diff
@@ -55,16 +58,52 @@
...
+# This function supports different reduction strategies such as computing the average
+# or maximum/minimum over the spot locations.
+# It also supports selecting a specific spot (TOP, MID, BOT) directly.
+# In this example we select the maximum value over all spots for each node and then request
+# the elemental location which implies averaging over all nodes in an element.
+# Using :func:`.get_ply_wise_data` has the advantage that all the averaging and filtering
+# is done on the server side.
```
Comment (PipKat) [inline examples/006_filter_composite_data_example.py:78]:
```suggestion
# This example selects the maximum value over all spots for each node and then requests
# the elemental location that implies averaging over all nodes in an element.
# Using the :func:`.get_ply_wise_data` function has the advantage that all the averaging and filtering
# is done on the server side.
```
Avoid we. Use "you" or rewrite with other subject in the active voice. (This example selects...)

File: examples/006_filter_composite_data_example.py [hunk 7]
```diff
@@ -55,16 +58,52 @@
...
+        component=Sym3x3TensorComponent.TENSOR11,
+        reduction_strategy=ReductionStrategy.MAX,
+        requested_location=dpf.locations.elemental,
+    )
+
+    composite_model.get_mesh().plot(elemental_max)
+
+
+# %%
+# Generic client side filtering
```
Comment (PipKat) [inline examples/006_filter_composite_data_example.py:93]:
```suggestion
# Generic client-side filtering
```

File: examples/006_filter_composite_data_example.py [hunk 8]
```diff
@@ -55,16 +58,52 @@
...
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+# This example shows how to filter data by layer, spot, and node using the generic filtering on
+# the client side.
+# Here we plot stress values in the material direction for the first node and top spot.
```
Comment (PipKat) [inline examples/006_filter_composite_data_example.py:97]:
```suggestion
# It plots stress values in the material direction for the first node and top spot.
```

File: examples/006_filter_composite_data_example.py [hunk 9]
```diff
@@ -82,23 +121,45 @@
...
 composite_model.get_mesh().plot(result_field)
 
+
 # %%
-# List analysis plies
-# ~~~~~~~~~~~~~~~~~~~
-# List all available analysis plies.
-all_ply_names = get_all_analysis_ply_names(composite_model.get_mesh())
-all_ply_names
+# Filtering by material
```
Comment (PipKat) [inline examples/006_filter_composite_data_example.py:126]:
```suggestion
# Filter by material
```
Use headings with simple verb forms, not gerunds, which are verbs with "-ing" endings.

File: examples/012_fatigue_example.py [hunk 10]
```diff
@@ -0,0 +1,155 @@
+"""
+.. _fatigue_plate_example:
+
+Evaluate fatigue for a composite plate
+--------------------------------------
+
+This example shows how to evaluate fatigue for a flat plate.
+It shows how PyPDF Composites can be used to select specific layers and define a custom
```
Comment (PipKat) [inline examples/012_fatigue_example.py:8]:
```suggestion
It shows how you can use PyPDF Composites to select specific layers and define a custom
```

File: examples/012_fatigue_example.py [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+combination method. For this example, the custom combination method is stress in fibre
+direction.
+
+A random load time series is created, and taking into account that the load is assumed
+proportional, rainflow counting is applied to load time series.
```
Comment (PipKat) [inline examples/012_fatigue_example.py:13]:
```suggestion
A random load time series is created. Taking into account that the load is assumed
proportional, rainflow counting is applied to the load time series.
```

File: examples/012_fatigue_example.py [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+Load ranges are then applied on the stress combination method and damage is evaluated
```
Comment (PipKat) [inline examples/012_fatigue_example.py:14]:
```suggestion
Load ranges are then applied on the stress combination method, and damage is evaluated
```

File: examples/012_fatigue_example.py [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+by using a dummy S-N curve.
+
+Be aware that the fatpack package is not developed by Ansys, so it is the responsibility
+of the user to verify that it works as expected. For further details:
+https://pypi.org/project/fatpack/
```
Comment (PipKat) [inline examples/012_fatigue_example.py:19]:
```suggestion
of the user to verify that it works as expected. For more information, see the `fatback package <https://pypi.org/project/fatpack/>`_,
```

File: examples/012_fatigue_example.py [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+# %%
+# Select Sigma11 as the combination method
+component = Sym3x3TensorComponent.TENSOR11
+
+
+# %%
+# Load time series and apply rainflow counting
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+# A random time series is created. Load is assumed proportional, so rainflow counting
+# can be directly done on load time series, to get load ranges.
```
Comment (PipKat) [inline examples/012_fatigue_example.py:79]:
```suggestion
# can be directly done on the load time series to get the load ranges.
```

File: examples/012_fatigue_example.py [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+plt.title("Dummy Linear S-N curve")
+plt.xlabel("Cycles to failure")
+plt.ylabel("Stress range (MPa)")
+
+
+# %%
+# Damage evaluation
+# ~~~~~~~~~~~~~~~~~
+# Stress S11 at time 1 and layer P1L1__ModelingPly.2 are read
+# for each load range and its damage is evaluated, using the dummy S-N curve
```
Comment (PipKat) [inline examples/012_fatigue_example.py:121]:
```suggestion
# for each load range. Its damage is evaluated using the dummy S-N curve.
```

File: src/ansys/dpf/composites/ply_wise_data.py [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+from typing import Union
+
+from ansys.dpf.core import Field, MeshedRegion, Operator, operators
+from ansys.dpf.gate.common import locations
+
+__all__ = ("ReductionStrategy", "get_ply_wise_data")
+
+
+class ReductionStrategy(Enum):
+    """The reduction strategy to get from spot values (BOT, MID, TOP) to a single value."""
```
Comment (PipKat) [inline src/ansys/dpf/composites/ply_wise_data.py:13]:
```suggestion
    """Provides the reduction strategy for getting from spot values (BOT, MID, TOP) to a single value."""
```

File: src/ansys/dpf/composites/ply_wise_data.py [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+
+    Parameters
+    ----------
+    field:
+        Field to extract data from.
+    ply_name:
+        Name of the ply to extract data from.
+    mesh :
+        Meshed region. Needs to be enriched with composite information.
+        Use the ``CompositeModel.get_mesh()`` method to get the meshed region.
```
Comment (PipKat) [inline src/ansys/dpf/composites/ply_wise_data.py:41]:
```suggestion
        Meshed region that needs to be enriched with composite information.
        Use the ``CompositeModel.get_mesh()`` method to get the meshed region.
```

**Comment by greschd** [comment]
@janvonrickenbach I changed the base branch to `feat/ply_wise_plots` to minimize the shown differences. It should automatically revert back to `main` once #396 is merged.

**Review by greschd — APPROVED** [review]

**Review by roosre — APPROVED** [review]

**Pull request merged**